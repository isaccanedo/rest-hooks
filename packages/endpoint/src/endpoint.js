function runCompat(endpoint, options) {
  endpoint.type = endpoint.sideEffect ? 'mutate' : 'read';
  endpoint.options = { ...options };
  delete endpoint.options.key;
  delete endpoint.options.schema;
  delete endpoint.options.sideEffect;
  delete endpoint.options.fetch;
  delete endpoint.options.getFetchKey;
  delete endpoint.options.options;
  if (Object.keys(endpoint.options).length === 0) {
    delete endpoint.options;
  }
  if (endpoint.schema === undefined) endpoint.schema = null;
}

let CSP = false;
try {
  Function();
} catch (e) {
  /* istanbul ignore next */
  CSP = true;
}

/**
 * Defines an async data source.
 * @see https://resthooks.io/docs/api/Endpoint
 */
export default class Endpoint extends Function {
  constructor(fetchFunction, options) {
    let self;
    // TODO: Test the fallback?
    /* istanbul ignore if */
    /* istanbul ignore next */
    if (CSP) {
      self = (...args) => self.fetch(...args);
      Object.setPrototypeOf(self, new.target.prototype);
    } else {
      super('return arguments.callee.fetch.apply(arguments.callee, arguments)');
      self = this;
    }
    /** The following is for compatibility with FetchShape */
    self.getFetchKey = params => self.key(params);

    if (fetchFunction) self.fetch = fetchFunction;

    if (options && 'name' in options) {
      self.__name = options.name;
      delete options.name;
    } else if (fetchFunction) {
      self.__name = fetchFunction.name;
    }
    Object.assign(self, options);
    Object.defineProperty(self, 'name', {
      get: function () {
        return this.__name;
      },
    });

    /** The following is for compatibility with FetchShape */
    runCompat(self, options);
    return self;
  }

  key(...args) {
    return `${this.name} ${JSON.stringify(args)}`;
  }

  bind(thisArg, ...args) {
    const fetchFunc = this.fetch;
    const keyFunc = this.key;
    return this.extend({
      fetch() {
        return fetchFunc.apply(thisArg ?? this, args);
      },
      key() {
        return keyFunc.apply(this, args);
      },
    });
  }

  extend(options) {
    // make a constructor/prototype based off this
    // extend from it and init with options sent
    class E extends this.constructor {}

    Object.assign(E.prototype, this);
    const instance = new E(options.fetch, options);

    /** The following is for compatibility with FetchShape */
    runCompat(instance, { ...this.options, ...options });

    return instance;
  }
}
export const ExtendableEndpoint = Endpoint;
