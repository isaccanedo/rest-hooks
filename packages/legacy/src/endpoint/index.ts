export type {
  FetchShape,
  ReadShape,
  MutateShape,
  DeleteShape,
} from './shapes.js';
export type {
  SetShapeParams,
  ParamsFromShape,
  OptimisticUpdateParams,
  SchemaFromShape,
  BodyFromShape,
  ReturnFromShape,
} from './types.js';
export { default as shapeToEndpoint } from './shapeToEndpoint.js';
