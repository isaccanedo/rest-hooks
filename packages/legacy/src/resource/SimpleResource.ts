/* istanbul ignore file */

import type {
  EndpointExtraOptions,
  AbstractInstanceType,
} from '@rest-hooks/endpoint';

import { ReadShape, MutateShape, DeleteShape } from '../rest-3/legacy.js';
import FlatEntity from './FlatEntity.js';
import Delete from './Delete.js';
import { SchemaDetail, Method, SchemaList } from './types.js';
import { NotImplementedError } from './errors.js';
import paramsToString from './paramsToString.js';

/** Represents an entity to be retrieved from a server.
 * Typically 1:1 with a url endpoint.
 *
 * This can be a useful organization for many REST-like API patterns.
 */
/** @deprecated in favor of @rest-hooks/rest */
export default abstract class SimpleResource extends FlatEntity {
  // typescript todo: require subclasses to implement
  /** Used as base of url construction */
  static readonly urlRoot: string;

  static toString() {
    return `${this.name}::${this.urlRoot}`;
  }

  /** Returns the globally unique identifier for this SimpleResource */
  static get key(): string {
    return this.urlRoot;
  }

  /** Warning level for probable malformed responses
   *
   * We only 'warn' here since this is the upgrade path
   */
  static automaticValidation = 'warn' as const;

  /** URL to find this SimpleResource */
  declare readonly url: string;

  private declare __url?: string;

  /** Get the url for a SimpleResource
   *
   * Default implementation conforms to common REST patterns
   */
  static url(urlParams: Readonly<Record<string, any>>): string {
    if (
      Object.hasOwn(urlParams, 'url') &&
      urlParams.url &&
      typeof urlParams.url === 'string'
    ) {
      return urlParams.url;
    }
    if (this.pk(urlParams as any) !== undefined) {
      return `${this.urlRoot.replace(/\/$/, '')}/${this.pk(urlParams as any)}`;
    }
    return this.urlRoot;
  }

  /** Get the url for many SimpleResources
   *
   * Default implementation conforms to common REST patterns
   */
  static listUrl(
    searchParams: Readonly<Record<string, string | number>> = {},
  ): string {
    if (Object.keys(searchParams).length) {
      return `${this.urlRoot}?${paramsToString(searchParams)}`;
    }
    return this.urlRoot;
  }

  /** Perform network request and resolve with json body */
  static fetch(
    method: Method,
    url: string,
    body?: Readonly<object | string>,
  ): Promise<any> {
    // typescript currently doesn't allow abstract static methods
    throw new NotImplementedError();
  }

  /** Perform network request and resolve with HTTP Response */
  static fetchResponse(
    method: Method,
    url: string,
    body?: Readonly<object | string>,
  ): Promise<Response> {
    // typescript currently doesn't allow abstract static methods
    throw new NotImplementedError();
  }

  /** @deprecated */
  /** Get the request options for this SimpleResource  */
  static getFetchOptions(): EndpointExtraOptions | undefined {
    return {
      errorPolicy: () => 'soft' as const,
    };
  }

  /** Init options for fetch - run at fetch */
  static getFetchInit(init: Readonly<RequestInit>): RequestInit {
    return init;
  }

  /** Get the request options for this SimpleResource  */
  static getEndpointExtra(): EndpointExtraOptions | undefined {
    return this.getFetchOptions();
  }

  // TODO: memoize these so they can be referentially compared
  /** Shape to get a single entity */
  static detailShape<T extends typeof SimpleResource>(
    this: T,
  ): ReadShape<SchemaDetail<Readonly<AbstractInstanceType<T>>>> {
    const getFetchKey = (params: Readonly<object>) => {
      return 'GET ' + this.url(params);
    };
    const options = this.getEndpointExtra();
    return {
      type: 'read',
      schema: this,
      options,
      getFetchKey,
      fetch: (params: Readonly<object>) => {
        return this.fetch('get', this.url(params));
      },
    };
  }

  /** Shape to get a list of entities */
  static listShape<T extends typeof SimpleResource>(
    this: T,
  ): ReadShape<SchemaList<Readonly<AbstractInstanceType<T>>>> {
    const getFetchKey = (params: Readonly<Record<string, string>>) => {
      return 'GET ' + this.listUrl(params);
    };
    const options = this.getEndpointExtra();
    return {
      type: 'read',
      schema: [this],
      options,
      getFetchKey,
      fetch: (params: Readonly<Record<string, string | number>>) => {
        return this.fetch('get', this.listUrl(params));
      },
    };
  }

  /** Shape to create a new entity (post) */
  static createShape<T extends typeof SimpleResource>(
    this: T,
  ): MutateShape<
    SchemaDetail<Readonly<AbstractInstanceType<T>>>,
    Readonly<object>,
    Partial<AbstractInstanceType<T>>
  > {
    const options = this.getEndpointExtra();
    return {
      type: 'mutate',
      schema: this,
      options,
      getFetchKey: (params: Readonly<Record<string, string>>) => {
        return 'POST ' + this.listUrl(params);
      },
      fetch: (
        params: Readonly<Record<string, string | number>>,
        body: Partial<AbstractInstanceType<T>>,
      ) => {
        return this.fetch('post', this.listUrl(params), body);
      },
    };
  }

  /** Shape to update an existing entity (put) */
  static updateShape<T extends typeof SimpleResource>(
    this: T,
  ): MutateShape<
    SchemaDetail<Readonly<AbstractInstanceType<T>>>,
    Readonly<object>,
    Partial<AbstractInstanceType<T>>
  > {
    const options = this.getEndpointExtra();
    return {
      type: 'mutate',
      schema: this,
      options,
      getFetchKey: (params: object) => {
        return 'PUT ' + this.url(params);
      },
      fetch: (
        params: Readonly<object>,
        body: Partial<AbstractInstanceType<T>>,
      ) => {
        return this.fetch('put', this.url(params), body);
      },
    };
  }

  /** Shape to update a subset of fields of an existing entity (patch) */
  static partialUpdateShape<T extends typeof SimpleResource>(
    this: T,
  ): MutateShape<
    SchemaDetail<Readonly<AbstractInstanceType<T>>>,
    Readonly<object>,
    Partial<AbstractInstanceType<T>>
  > {
    const options = this.getEndpointExtra();
    return {
      type: 'mutate',
      schema: this,
      options,
      getFetchKey: (params: Readonly<object>) => {
        return 'PATCH ' + this.url(params);
      },
      fetch: (
        params: Readonly<object>,
        body: Partial<AbstractInstanceType<T>>,
      ) => {
        return this.fetch('patch', this.url(params), body);
      },
    };
  }

  /** Shape to delete an entity (delete) */
  static deleteShape<T extends typeof SimpleResource>(
    this: T,
  ): DeleteShape<Delete<T>, Readonly<object>> {
    const options = this.getEndpointExtra();
    return {
      type: 'mutate',
      schema: new Delete(this),
      options,
      getFetchKey: (params: object) => {
        return 'DELETE ' + this.url(params);
      },
      fetch: (params: Readonly<object>) => {
        return this.fetch('delete', this.url(params)).then(() => params);
      },
    };
  }
}

// We're only allowing this to get set for descendants but
// by default we want Typescript to treat it as readonly.
Object.defineProperty(SimpleResource.prototype, 'url', {
  get(): string {
    if (this.__url !== undefined) return this.__url;
    // typescript thinks constructor is just a function
    const Static = this.constructor as typeof SimpleResource;
    return Static.url(this);
  },
  set(url: string) {
    this.__url = url;
  },
});
