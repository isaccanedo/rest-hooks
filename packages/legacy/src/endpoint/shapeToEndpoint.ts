import { Endpoint } from '@rest-hooks/endpoint';
import type { EndpointInstance } from '@rest-hooks/endpoint';

import type { FetchShape } from './shapes.js';

type ShapeTypeToSideEffect<T extends 'read' | 'mutate' | 'delete' | undefined> =
  T extends 'read' | undefined ? undefined : true;

const SIDEEFFECT_TYPES: (string | undefined)[] = ['mutate', 'delete'];

export default function shapeToEndpoint<
  Shape extends Partial<FetchShape<any, any>>,
>(
  shape: Shape,
): Shape['fetch'] extends (...args: any) => Promise<any>
  ? EndpointInstance<
      Shape['fetch'],
      Shape['schema'],
      ShapeTypeToSideEffect<Shape['type']>
    > &
      Shape['options']
  : Shape['options'] & { key: Shape['getFetchKey'] } {
  // make this identity function for endpoints
  if ('key' in shape) return shape as any;
  const sideEffect = SIDEEFFECT_TYPES.includes(shape.type);
  const options = {
    ...shape.options,
    key: shape.getFetchKey,
    schema: shape.schema,
    ...((sideEffect && { sideEffect }) as any),
  };
  if (Object.hasOwn(shape, 'fetch'))
    return new Endpoint(shape.fetch as any, options);

  return options as any;
}
