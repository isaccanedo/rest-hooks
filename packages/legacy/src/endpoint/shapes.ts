import { Schema, EndpointExtraOptions } from '@rest-hooks/endpoint';

/** Defines the shape of a network request */
export interface FetchShape<
  S extends Schema | undefined,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object | string> | void | unknown =
    | Readonly<object | string>
    | undefined,
  Response = any,
> {
  readonly type: 'read' | 'mutate' | 'delete';
  fetch(params: Params, body?: Body): Promise<Response>;
  getFetchKey(params: Params): string;
  readonly schema: S;
  readonly options?: EndpointExtraOptions;
}

/** To change values on the server */
export interface MutateShape<
  S extends Schema | undefined,
  Params extends Readonly<object> = Readonly<object>,
  Body extends Readonly<object | string> | void | unknown =
    | Readonly<object | string>
    | undefined,
  Response extends object | string | number | boolean | null = any,
> extends FetchShape<S, Params, Body, Response> {
  readonly type: 'mutate';
  fetch(params: Params, body: Body): Promise<Response>;
}

/** Removes entities */
export interface DeleteShape<
  S extends Schema | undefined,
  Params extends Readonly<object> = Readonly<object>,
  Response extends object | string | number | boolean | null = any,
> extends FetchShape<S, Params, undefined, Response> {
  readonly type: 'mutate';
  fetch(params: Params, ...args: any): Promise<Response>;
}

/** For retrieval requests */
export interface ReadShape<
  S extends Schema | undefined,
  Params extends Readonly<object> = Readonly<object>,
  Response extends object | string | number | boolean | null = any,
> extends FetchShape<S, Params, undefined, Response> {
  readonly type: 'read';
  fetch(params: Params): Promise<Response>;
}
