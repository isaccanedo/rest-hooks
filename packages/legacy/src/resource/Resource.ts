/* istanbul ignore file */

import SimpleResource from './SimpleResource.js';
import type { Method } from './types.js';

class NetworkError extends Error {
  declare status: number;
  declare response: Response;
  name = 'NetworkError';

  constructor(response: Response) {
    super(
      response.statusText || `Network response not 'ok': ${response.status}`,
    );
    this.status = response.status;
    this.response = response;
  }
}

/**
 * Represents an entity to be retrieved from a server.
 * Typically 1:1 with a url endpoint.
 */
/** @deprecated in favor of @rest-hooks/rest */
export default abstract class Resource extends SimpleResource {
  /** A function to mutate all request options for fetch */
  static fetchOptionsPlugin?: (options: RequestInit) => RequestInit;

  /** Init options for fetch - run at fetch */
  static getFetchInit(init: Readonly<RequestInit>): RequestInit {
    if (this.fetchOptionsPlugin) return this.fetchOptionsPlugin(init);
    return init;
  }

  /** Perform network request and resolve with HTTP Response */
  static fetchResponse(
    method: Method,
    url: string,
    body?: Readonly<object | string>,
  ) {
    let options: RequestInit = {
      method: method.toUpperCase(),
    };
    options = this.getFetchInit(options);
    if (body && isPojo(body)) options.body = JSON.stringify(body);
    else if (body) options.body = body;
    if (!options.body || typeof options.body === 'string') {
      options.headers = {
        'Content-Type': 'application/json',
        ...options.headers,
      };
    }
    return fetch(url, options)
      .then(response => {
        if (!response.ok) {
          throw new NetworkError(response);
        }
        return response;
      })
      .catch(error => {
        // ensure CORS, network down, and parse errors are still caught by NetworkErrorBoundary
        if (error instanceof TypeError) {
          (error as any).status = 400;
        }
        throw error;
      });
  }

  /** Perform network request and resolve with json body */
  static fetch(method: Method, url: string, body?: Readonly<object | string>) {
    return this.fetchResponse(method, url, body).then((response: Response) => {
      if (
        !response.headers.get('content-type')?.includes('json') ||
        response.status === 204
      )
        return response.text();
      return response.json().catch(error => {
        error.status = 400;
        throw error;
      });
    });
  }
}

const proto = Object.prototype;
const gpo = Object.getPrototypeOf;

function isPojo(obj: unknown): obj is Record<string, any> {
  if (obj === null || typeof obj !== 'object') {
    return false;
  }
  return gpo(obj) === proto;
}
