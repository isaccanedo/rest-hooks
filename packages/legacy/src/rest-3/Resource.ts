import SimpleResource from './SimpleResource.js';

class NetworkError extends Error {
  declare status: number;
  declare response: Response;
  name = 'NetworkError';

  constructor(response: Response) {
    super(
      response.statusText ||
        /* istanbul ignore next */ `Network response not 'ok': ${response.status}`,
    );
    this.status = response.status;
    this.response = response;
  }
}

/**
 * Represents an entity to be retrieved from a server.
 * Typically 1:1 with a url endpoint.
 * @see https://resthooks.io/docs/api/resource
 */
export default abstract class Resource extends SimpleResource {
  /** Perform network request and resolve with HTTP Response */
  static fetchResponse(input: RequestInfo, init: RequestInit) {
    let options: RequestInit = init;
    if (!options.body || typeof options.body === 'string') {
      options = {
        ...options,
        headers: {
          'Content-Type': 'application/json',
          ...options.headers,
        },
      };
    }
    return fetch(input, options)
      .then(response => {
        if (!response.ok) {
          throw new NetworkError(response);
        }
        return response;
      })
      .catch(error => {
        // ensure CORS, network down, and parse errors are still caught by NetworkErrorBoundary
        if (error instanceof TypeError) {
          (error as any).status = 400;
        }
        throw error;
      });
  }

  /** Perform network request and resolve with json body */
  static fetch(input: RequestInfo, init: RequestInit) {
    return this.fetchResponse(input, init).then((response: Response) => {
      if (
        !response.headers.get('content-type')?.includes('json') ||
        response.status === 204
      )
        return response.text();
      return response.json().catch(error => {
        error.status = 400;
        throw error;
      });
    });
  }
}
