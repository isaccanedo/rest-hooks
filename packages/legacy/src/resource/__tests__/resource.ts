import {
  CoolerArticleResource,
  UserResource,
  UrlArticleResource,
} from '__tests__/legacy';
import nock from 'nock';
import { normalize } from '@rest-hooks/normalizr';
import { schema } from '@rest-hooks/endpoint';

import Resource from '../Resource';
import SimpleResource from '../SimpleResource';

function onError(e: any) {
  e.preventDefault();
}
beforeEach(() => {
  if (typeof addEventListener === 'function')
    addEventListener('error', onError);
});
afterEach(() => {
  if (typeof removeEventListener === 'function')
    removeEventListener('error', onError);
});

describe('Resource', () => {
  beforeAll(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200);
  });

  afterAll(() => {
    nock.cleanAll();
  });

  it('should implement schema.EntityInterface', () => {
    class A extends Resource {
      readonly id: string = '';
      pk() {
        return this.id;
      }
    }
    const a: schema.EntityInterface = A;
  });

  it('should init', () => {
    const resource = CoolerArticleResource.fromJS({
      id: 5,
      title: 'happy',
      author: 5,
    });
    expect(resource.pk()).toBe('5');
    expect(CoolerArticleResource.pk(resource)).toBe('5');
    expect(resource.title).toBe('happy');
    expect(resource.things).toBe('happy five');
    expect(resource.url).toBe('http://test.com/article-cooler/5');
    expect(resource.author).toBe(5);
  });

  it('should not init Resource itself', () => {
    expect(() => Resource.fromJS({})).toThrow();
  });

  it('should work with `url` member', () => {
    expect(() => UrlArticleResource.fromJS({})).not.toThrow();
    expect(() => UrlArticleResource.fromJS({ url: 'five' })).not.toThrow();
    const urlArticle = UrlArticleResource.fromJS({ url: 'five' });
    expect(urlArticle.url).toBe('five');
  });

  it('should convert class to string', () => {
    expect(CoolerArticleResource.toString()).toBeDefined();
    expect(CoolerArticleResource.toString()).toMatchInlineSnapshot(
      `"CoolerArticleResource::http://test.com/article-cooler/"`,
    );
  });

  it('should render url property', () => {
    const article = CoolerArticleResource.fromJS({
      id: 5,
      title: 'happy',
      author: 5,
    });
    expect(article.url).toBe('http://test.com/article-cooler/5');
  });

  it('should render url property with non-trailing slash root', () => {
    // Remove trailing slash in url root
    CoolerArticleResource.urlRoot = 'http://test.com/article-cooler';
    const article = CoolerArticleResource.fromJS({
      id: 5,
      title: 'happy',
      author: 5,
    });
    expect(article.url).toBe('http://test.com/article-cooler/5');
    // Reset for future tests
    CoolerArticleResource.urlRoot = 'http://test.com/article-cooler/';
  });

  const [coolA, coolB, coolC] = [
    CoolerArticleResource.fromJS({ title: 'great' }),
    CoolerArticleResource.fromJS({ id: 5 }),
    CoolerArticleResource.fromJS({}),
  ];

  describe('merge()', () => {
    const c = CoolerArticleResource.merge(coolA, coolB);
    it('works with partial', () => {
      expect(c.things).toBeDefined();
      expect(c.id).toBe(5);
      expect(c.content).toBe('');
      expect(c.title).toBe('great');
      expect(c).toMatchInlineSnapshot(`
        CoolerArticleResource {
          "author": null,
          "content": "",
          "id": 5,
          "tags": [],
          "title": "great",
        }
      `);
    });
    it('works with definedObjects()', () => {
      expect(c).toMatchInlineSnapshot(`
        CoolerArticleResource {
          "author": null,
          "content": "",
          "id": 5,
          "tags": [],
          "title": "great",
        }
      `);
    });
    it('does nothing with empty arg', () => {
      expect(CoolerArticleResource.merge(c, coolC)).toEqual(c);
    });
  });

  describe('hasDefined()', () => {
    it('works ', () => {
      expect(CoolerArticleResource.hasDefined(coolA, 'title')).toBe(true);
      expect(CoolerArticleResource.hasDefined(coolA, 'author')).toBe(false);
    });
  });

  describe('toObjectDefined()', () => {
    it('works', () => {
      expect(CoolerArticleResource.toObjectDefined(coolA))
        .toMatchInlineSnapshot(`
        {
          "title": "great",
        }
      `);
      expect(CoolerArticleResource.toObjectDefined(coolB))
        .toMatchInlineSnapshot(`
        {
          "id": 5,
        }
      `);
    });
  });

  describe('keysDefined()', () => {
    it('works', () => {
      expect(CoolerArticleResource.keysDefined(coolA)).toMatchInlineSnapshot(`
        [
          "title",
        ]
      `);
      expect(CoolerArticleResource.keysDefined(coolB)).toMatchInlineSnapshot(`
        [
          "id",
        ]
      `);
    });
  });

  describe('listUrl', () => {
    it('should listUrl with an arg', () => {
      expect(CoolerArticleResource.listUrl({ author: 5 })).toBe(
        'http://test.com/article-cooler/?author=5',
      );
    });
    it('should listUrl with no args', () => {
      expect(CoolerArticleResource.listUrl({})).toBe(
        'http://test.com/article-cooler/',
      );
    });
    it('should sort consistently', () => {
      expect(
        CoolerArticleResource.listUrl({
          z: 'alpha',
          y: 'beta',
          m: 'never',
          a: 'sometimes',
          c: 'again',
        }),
      ).toBe(
        'http://test.com/article-cooler/?a=sometimes&c=again&m=never&y=beta&z=alpha',
      );
    });
  });

  describe('static url', () => {
    it('should use provided url parameter as value', () => {
      const url = 'this is the url';
      expect(UserResource.url({ url })).toBe(url);
    });
    it('should return root if cannot find params with url', () => {
      expect(UserResource.url({})).toMatchInlineSnapshot(
        `"http://test.com/user/"`,
      );
    });
  });

  describe('Resource.fetch()', () => {
    const id = 5;
    const idHtml = 6;
    const idNoContent = 7;
    const payload = {
      id,
      title: 'happy',
      author: 5,
    };
    const putResponseBody = {
      id,
      title: 'happy',
      completed: true,
    };
    const patchPayload = {
      title: 'happy',
    };
    const patchResponseBody = {
      id,
      title: 'happy',
      completed: false,
    };

    beforeEach(() => {
      nock(/.*/)
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        })
        .get(`/article-cooler/${payload.id}`)
        .reply(200, payload)
        .get(`/article-cooler/${idHtml}`)
        .reply(200, '<body>this is html</body>')
        .get(`/article-cooler/${idNoContent}`)
        .reply(204, '')
        .post('/article-cooler/')
        .reply((uri, requestBody) => [
          201,
          requestBody,
          { 'content-type': 'application/json' },
        ])
        .put('/article-cooler/5')
        .reply((uri, requestBody) => {
          let body = requestBody as any;
          if (typeof requestBody === 'string') {
            body = JSON.parse(requestBody);
          }
          for (const key of Object.keys(CoolerArticleResource.fromJS({}))) {
            if (key !== 'id' && !(key in body)) {
              return [400, {}, { 'content-type': 'application/json' }];
            }
          }
          return [200, putResponseBody, { 'content-type': 'application/json' }];
        })
        .patch('/article-cooler/5')
        .reply(() => [
          200,
          patchResponseBody,
          { 'content-type': 'application/json' },
        ])
        .intercept('/article-cooler/5', 'DELETE')
        .reply(200, {});
    });

    it('should throw with SimpleResource', () => {
      expect(() =>
        SimpleResource.fetch(
          'get',
          CoolerArticleResource.url({
            id: payload.id,
          }),
        ),
      ).toThrow();
    });

    it('fetchResponse() should throw with SimpleResource', () => {
      expect(() =>
        SimpleResource.fetchResponse(
          'get',
          CoolerArticleResource.url({
            id: payload.id,
          }),
        ),
      ).toThrow();
    });

    it('should GET', async () => {
      const article = await CoolerArticleResource.fetch(
        'get',
        CoolerArticleResource.url({
          id: payload.id,
        }),
      );
      expect(article).toBeDefined();
      if (!article) {
        throw new Error('ahh');
      }
      expect(article.title).toBe(payload.title);
    });

    it('should POST', async () => {
      const payload2 = { id: 20, content: 'better task' };
      const article = await CoolerArticleResource.fetch(
        'post',
        CoolerArticleResource.listUrl(),
        JSON.stringify(payload2),
      );
      expect(article).toMatchObject(payload2);
    });

    it('should PUT with multipart form data', async () => {
      const payload2 = { id: 500, content: 'another' };
      let lastRequest: any;
      nock(/.*/)
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
        })
        .put('/article-cooler/500')
        .reply(function (uri, requestBody) {
          lastRequest = this.req;
          return [201, payload2, { 'Content-Type': 'application/json' }];
        });
      const newPhoto = new Blob();
      const body = new FormData();
      body.append('photo', newPhoto);

      const article = await CoolerArticleResource.fetch(
        'put',
        CoolerArticleResource.url({ id: '500' }),
        body,
      );
      expect(lastRequest.headers['content-type']).toContain(
        'multipart/form-data',
      );
      expect(article).toMatchObject(payload2);
    });

    it('should DELETE', async () => {
      const res = await CoolerArticleResource.fetch(
        'delete',
        CoolerArticleResource.url({
          id: payload.id,
        }),
      );
      expect(res).toEqual({});
    });

    it('should PUT', async () => {
      const response = await CoolerArticleResource.fetch(
        'put',
        CoolerArticleResource.url(payload),
        { ...CoolerArticleResource.fromJS(payload) },
      );
      expect(response).toEqual(putResponseBody);
    });

    it('should PATCH', async () => {
      const response = await CoolerArticleResource.fetch(
        'patch',
        CoolerArticleResource.url({ id }),
        JSON.stringify(patchPayload),
      );
      expect(response).toEqual(patchResponseBody);
    });

    it('should throw if response is not json', async () => {
      let error: any;
      try {
        await CoolerArticleResource.fetch(
          'get',
          CoolerArticleResource.url({ id: idHtml }),
        );
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      // This is very weird, but we're forced to use node-fetch for react native
      // node-fetch doesn't handle errors consistently with normal fetch implementations, so this won't work
      // react-native itself should match this correctly however.
      if (typeof window !== 'undefined') expect(error.status).toBe(400);
    });

    it('should throw if network is down', async () => {
      const oldError = console.error;
      console.error = () => {};

      const id = 10;
      nock(/.*/)
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        })
        .get(`/article-cooler/${id}`)
        .replyWithError(new TypeError('Network Down'));

      let error: any;
      try {
        await CoolerArticleResource.fetch(
          'get',
          CoolerArticleResource.url({ id }),
        );
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.status).toBe(400);

      // eslint-disable-next-line require-atomic-updates
      console.error = oldError;
    });

    it('should return raw response if status is 204 No Content', async () => {
      const res = await CoolerArticleResource.fetch(
        'get',
        CoolerArticleResource.url({ id: idNoContent }),
      );
      expect(res).toBe('');
    });

    it('should return text if content-type is not json', async () => {
      const id = 8;
      const text = '<body>this is html</body>';
      nock(/.*/)
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        })
        .get(`/article-cooler/${id}`)
        .reply(200, text, { 'content-type': 'html/text' });

      const res = await CoolerArticleResource.fetch(
        'get',
        CoolerArticleResource.url({ id }),
      );
      expect(res).toBe('<body>this is html</body>');
    });

    it('should return text if content-type does not exist', async () => {
      const id = 10;
      const text = '<body>this is html</body>';
      nock(/.*/)
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
        })
        .get(`/article-cooler/${id}`)
        .reply(200, text, {});

      const res = await CoolerArticleResource.fetch(
        'get',
        CoolerArticleResource.url({ id }),
      );
      expect(res).toBe(text);
    });
  });

  describe('nested schema', () => {
    describe('merging', () => {
      const nested = [
        {
          id: 5,
          title: 'hi ho',
          content: 'whatever',
          tags: ['a', 'best', 'react'],
          author: {
            id: 23,
            username: 'bob',
            email: 'bob@bob.com',
          },
        },
        {
          id: 3,
          title: 'the next time',
          content: 'whatever',
          author: {
            id: 23,
            username: 'charles',
          },
        },
      ];
      const { schema } = CoolerArticleResource.listWithUserRequest();
      const normalized = normalize(nested, schema);

      // TODO: fix normalize types so we know this is actuaally multiple things
      const user: any = normalized.entities[UserResource.key]['23'];
      it('should include nested user', () => {
        expect(user).toBeDefined();
      });
      it('should have user be UserResource type', () => {
        expect(user).toBeInstanceOf(UserResource);
      });
      it('should take property of first if not set in second', () => {
        expect(user.email).toBe('bob@bob.com');
      });
      it('should overwrite second properties over first', () => {
        expect(user.username).toBe('charles');
      });
      it('should use default for unset propreties', () => {
        expect(UserResource.hasDefined(user, 'isAdmin')).toBe(false);
      });
      it('should match snapshot', () => {
        expect(normalized).toMatchSnapshot({ entityMeta: expect.any(Object) });
      });
    });

    it('should throw a custom error if data does not include pk', () => {
      const schema = CoolerArticleResource;
      function normalizeBad() {
        normalize({ content: 'hi' }, schema);
      }
      expect(normalizeBad).toThrowErrorMatchingSnapshot();
    });
  });
});
