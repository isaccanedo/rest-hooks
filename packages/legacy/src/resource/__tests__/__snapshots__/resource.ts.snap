// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource nested schema merging should match snapshot 1`] = `
{
  "entities": {
    "http://test.com/article-cooler/": {
      "3": CoolerArticleResource {
        "author": "23",
        "content": "whatever",
        "id": 3,
        "tags": [],
        "title": "the next time",
      },
      "5": CoolerArticleResource {
        "author": "23",
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
    },
    "http://test.com/user/": {
      "23": UserResource {
        "email": "bob@bob.com",
        "id": 23,
        "isAdmin": false,
        "username": "charles",
      },
    },
  },
  "entityMeta": Any<Object>,
  "indexes": {},
  "result": [
    "5",
    "3",
  ],
}
`;

exports[`Resource nested schema should throw a custom error if data does not include pk 1`] = `
"Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema

  Entity: CoolerArticleResource
  Value: {
  "content": "hi"
}
  "
`;
